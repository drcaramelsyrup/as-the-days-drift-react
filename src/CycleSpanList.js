import React from 'react';
import CycleSpan from './CycleSpan';
import { matchesQuery } from './InventoryUtils';
import { makeCycle, makeCycleData, updatedInventoryForCycle } from './CycleUtils';
import './CycleSpanList.css';

const CycleSpanList = (props) => {
	return <div className='CycleSpanList'>{
		props.data != null && getCycleSpans(
			props.data, 
			props.inventory,
			props.callback)	
	}</div>;
}

const getCycleSpans = (data, inventory = {}, callback = null) => {

  // TODO: bad practice to use idx in a component array
  const cycleArray = preprocessData(data, inventory);
  return cycleArray.reduce((acc, cycle, idx) => {
	return acc.concat([ <CycleSpan 
		key={ 'cyclespan' + idx } 
		cycle={ cycle }
		callback={ cycleSpanUpdateFunction(cycle, inventory, callback) }
	/> ]);
  }, []);

}

const resolveConditionals = (rawData, conditionals, inventory) => {

	if (conditionals == null)
		return rawData;

	const ids = conditionals.reduce((acc, conditional) => {
		if (matchesQuery(conditional.conditions, inventory))
			acc.relevant = acc.relevant.concat([ conditional.id ]);
		else
			acc.irrelevant = acc.irrelevant.concat([ conditional.id ]);
		return acc;
	}, { relevant: [], irrelevant: [] });

	const resolveTextWithConditional = (inText, conditionalId, conditionalText) => {
		if (conditionalId == null)
			return inText;
		if (!isValidString(conditionalText))
			return inText.replace(conditionalId, '');
		return inText.replace(conditionalId, conditionalText);
	}

	const resolvedData = Object.assign({}, rawData);

	const irrelevantIds = ids.irrelevant;
	resolvedData.text = irrelevantIds.reduce((text, id) => {
		return resolveTextWithConditional(text, id, '');
	}, resolvedData.text);

	const relevantIds = ids.relevant;
	resolvedData.text = relevantIds.reduce((text, id) => {
		// TODO: map this to optimize instead of running O(n) find by predicate?
		const conditionalIdx = conditionals.findIndex((cond) => { return cond.id === id; });
		const conditionalText = conditionalIdx >= 0
			? conditionals[conditionalIdx].text
			: '';
		return resolveTextWithConditional(text, id, conditionalText);
	}, resolvedData.text);

	return resolvedData;

}

const preprocessData = (rawData, inventory) => {
	
	const jsonData = resolveConditionals(rawData, rawData.conditionals, inventory);
	if (!isValidString(jsonData.text))
		return [];

	// Filter out cycles that do not appear and sort by order of appearance.
	const cycleIds = jsonData.cycles != null
		? Object.keys(jsonData.cycles)
			.filter(id => jsonData.text.indexOf(id) >= 0)
			.sort((a, b) => {
				return jsonData.text.indexOf(a) > jsonData.text.indexOf(b);
			})
		: [];

	// Only pure text, no cycles.
	if (cycleIds.length <= 0)
		return [ makeCycle(null, null, makeCycleData(jsonData.text)) ];

	const isValidIndex = (idx) => {
		return idx >= 0 && idx < cycleIds.length;
	}

	return cycleIds.reduce((acc, id, idx) => {

		// Start at the beginning or at the last cycle id position.
		const start = isValidIndex(idx - 1) 
			? jsonData.text.indexOf(cycleIds[idx - 1]) + cycleIds[idx - 1].length
			: 0;
		const end = isValidIndex(idx + 1)
			? jsonData.text.indexOf(cycleIds[idx + 1])
			: jsonData.text.length;

		const position = jsonData.text.indexOf(id);

		// Print before and after if it's the first, else print the after.
		const isFirstId = !isValidIndex(idx - 1);
		const before = isFirstId
			? jsonData.text.substr(start, position - start)
			: '';

		const after = jsonData.text.substr(
			position + id.length, end - position - id.length);

		const toConcat = [];

		if (before.length > 0)
			toConcat.push(makeCycle(null, null, makeCycleData(before)));

		const cycleData = jsonData.cycles[id];
		toConcat.push(makeCycle(id, getCycleIdx(id, cycleData, inventory), ...cycleData));

		if (after.length > 0)
			toConcat.push(makeCycle(null, null, makeCycleData(after)));

		return acc.concat(toConcat);

	}, []);
}

const getCycleIdx = (cycleId, cycleData, inventory) => {
	const desiredIdx = inventory.cycles != null ? inventory.cycles[cycleId] : 0;

	// Make sure that we have a valid cycle index.
	return getValidCycleIdx(cycleId, desiredIdx, cycleData, inventory);
}

const nextValidCycleIdx = (cycleId, currentIdx, cycleData, inventory) => {
	return getValidCycleIdx(cycleId, currentIdx + 1, cycleData, inventory);
}

const getValidCycleIdx = (cycleId, desiredIdx, cycleData, inventory) => {
	// Starting from the desired idx...
	if (desiredIdx == null)
		desiredIdx = 0;
	for (let i = desiredIdx; i < cycleData.length; ++i) {
		const datum = cycleData[i];
		if (matchesQuery(datum.conditions, inventory))
			return i;
	}
	for (let j = 0; j < desiredIdx; ++j) {
		const datum = cycleData[j];
		if (matchesQuery(datum.conditions, inventory))
			return j;
	}
	return -1;
}

// Returns false if string is empty, null, or undefined.
const isValidString = (str) => {
	return str && str.length > 0;
}

const cycleSpanUpdateFunction = (cycle, inventory, callback) => {
	return () => {
		const nextCycleIdx = nextValidCycleIdx(
			cycle.cycle_id, cycle.cycle_idx, cycle.data, inventory);
		if (nextCycleIdx === cycle.cycle_idx)
			return;

		callback(updatedInventoryForCycle(
			cycle.cycle_id, nextCycleIdx, cycle.data, inventory));
	}
}

export default CycleSpanList;
export { cycleSpanUpdateFunction };
