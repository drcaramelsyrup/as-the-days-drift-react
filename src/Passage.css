.Passage {
	position: absolute;
	z-index: 1;
	width: 80%;
	height: 60%;

	padding: 10%;
	line-height: 1.5;
	font-size: medium;
}