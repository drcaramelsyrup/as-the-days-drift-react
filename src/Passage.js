import React from 'react';
import CycleSpanList from './CycleSpanList';

const Passage = (props) => {
  return <div>{
    props.data != null && createCycleSpanList(
      props.data.pid, 
      props.data.actions, 
      props.data.conditionals, 
      props.data.cycles, 
      props.data.text)
  }</div>;
}

const createCycleSpanList = (id, actions, conditionals, cycles, text) => {

  const wrapper = (input) => {
    return input == null ? null : input;
  }

  return <CycleSpanList data={
    {
      actions: wrapper(actions),
      conditionals: wrapper(conditionals),
      cycles: wrapper(cycles),
      id: wrapper(id),
      text: wrapper(text)
    }
  }/>

}

export default Passage;
