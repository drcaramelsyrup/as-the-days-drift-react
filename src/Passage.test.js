import React from 'react';
import ReactDOM from 'react-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Passage from './Passage';
import CycleSpanList from './CycleSpanList';
import snapshotData from './test_passage.json';

describe('Passage - Smoke', () => {

	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<Passage />, div);
		ReactDOM.unmountComponentAtNode(div);
	});

});

const makePassageData = (data) => {
	const newData = Object.assign({}, data);
	newData.pid = newData.id;
	return newData;
}

describe('Passage - Render', () => {

	const makeExpectedListData = (id, actions, conditionals, cycles, text) => {
		return {
			actions: actions,
			conditionals: conditionals,
			cycles: cycles,
			id: id,
			text: text
		};	
	}

	const makeExpectedCycleSpanList = (data, inventory, callback) => {
		return <CycleSpanList 
			data={ data } 
			inventory={ inventory }
			callback={ callback } />;
	}

	it('does not render if not given data', () => {
		const renderer = new ShallowRenderer();

		renderer.render(<Passage />);
		const output = renderer.getRenderOutput();
		expect(output).toBe(false);

		renderer.unmount();
	});

	it('renders as a div container', () => {
		const renderer = new ShallowRenderer();

		const dummyPassage = {
			pid: 215,
			text: 'test text' 
		};
		renderer.render(<Passage data={ dummyPassage }/>);
		const output = renderer.getRenderOutput();
		expect(output.type).toBe('div');

		renderer.unmount();
	});

	it('renders with a simple passage id and text', () => {
		let text = 'test text';
		const pid = 215; 
		const dummyPassage = {
			pid: pid,
			text: text 
		};

		const passage = shallow(<Passage data={ dummyPassage } />);
		expect(passage.props().children).toEqual(
			makeExpectedCycleSpanList(
				makeExpectedListData(pid, null, null, null, text),
				{},
				passage.instance().updatePassage));
	});

	it('renders and passes along data fields', () => {
		const inventory = { trinkets: 2 };
		const pid = 0;
		const data = {
			actions: {
				learned: 2
			},
			conditionals: {
				conditions: {
					partner: {
						type: 'is',
						val: 'sam'
					}
				},
				id: '_cond_text_0',
				text: 'conditional '
			},
			cycles: {
				_cycle_example: [
					{
						actions: null,
						text: 'test cycle'
					},
					{
						actions: {
							career: 'journalist',
						},
						text: 'test cycle 2'
					}
				]
			},
			text: 'test text',
			id: pid
		};

		const wrapper = shallow(<Passage 
			data={ makePassageData(data) } 
			inventory={ inventory} />);
		expect(wrapper.props().children).toEqual(
			makeExpectedCycleSpanList(
				data, 
				inventory,
				wrapper.instance().updatePassage));
	});

	it('renders with a complete data passage', () => {
		const tree = TestRenderer.create(<Passage data={ snapshotData } />).toJSON();
		expect(tree).toMatchSnapshot();
	});

});

describe('Passage - updatePassage', () => {

	const pid = 42;
	const text = 'update passage ';
	const dummyPassage = {
		pid: pid,
		text: text
	};

	it('renders and updates its inventory state', () => {
		const inventory = { conventional: 1 };
		const passage = shallow(<Passage 
			data={ dummyPassage }
			inventory={ inventory } />);
		const newInventory = { unconventional: 1 };
		passage.instance().updatePassage(newInventory);
		expect(passage.state().inventory).toEqual(newInventory);
	});

	it('renders and passes the update callback to children', () => {
		const inventory = { conventional: 1, practical: 2 };
		const passage = shallow(<Passage 
			data={ dummyPassage }
			inventory={ inventory } />);

		const newInventory = { unconventional: 1, romantic: 3 };
		passage.instance().updatePassage(newInventory);
		expect(passage.state().inventory).toEqual(newInventory);
	});

});

describe('Passage - advancePassage', () => {

	const data = { id: 42, text: 'test' };
	const passageData = makePassageData(data);
	const newData = { id: 45, text: 'new text' };
	const newPassageData = makePassageData(newData);

	it('renders and updates its passage data', () => {
		const inventory = { practical: 3 };

		const passage = shallow(<Passage
			data={ passageData }
			inventory={ inventory } />);
		expect(passage.state().data).toEqual(passageData);

		const newInventory = { unconventional: 2 };

		passage.instance().advancePassage(newPassageData, newInventory);
		expect(passage.state().data).toEqual(newPassageData);
		expect(passage.state().inventory).toEqual(newInventory);
	});

	it('renders and integrates inventory actions', () => {
		const inventory = { 
			actions: {
				romantic: 1,
				neurotic: 1
			},
			romantic: 2
		};
		const integratedInventory = {
			romantic: 3,
			neurotic: 1
		};

		const passage = shallow(<Passage
			data={ passageData }
			inventory={ inventory } />);

		passage.instance().advancePassage(newPassageData, inventory);
		expect(passage.state().data).toEqual(newPassageData);
		expect(passage.state().inventory).toEqual(integratedInventory);
	});

});

