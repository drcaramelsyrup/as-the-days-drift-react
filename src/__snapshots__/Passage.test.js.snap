// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passage - Render renders with a complete data passage 1`] = `
<div>
  <div>
    <span
      className="cycle"
    />
    <span
      className="cycle"
    />
    <span
      className="cycle"
    />
    <span
      className="cycle"
    />
    <span
      className="cycle"
    />
  </div>
</div>
`;
